### PR TITLE
BUG: Fix bugs with running and add code coverage

### LINKED ISSUE #55: use pytest for tests
As we discussed [here](https://github.com/mne-tools/mne-study-template/pull/53#issuecomment-524610386), we should pytest.parametrize and as much of the pytest infrastructure as possible rather than reinventing the wheel.

### LINKED ISSUE #654: on_error='debug' is somehow behaving incorrectly / is broken
I edited `config_ERP_CORE.py` to contain the following settings:

```python
on_error = 'debug'
# parallel_backend = 'dask'
```

Then I edited `_02_frequency_filter.py` to raise an exception:
```python
@failsafe_run(script_path=__file__,
              get_input_fnames=get_input_fnames_frequency_filter)
def filter_data(
    *,
    cfg,
    subject: str,
    session: Optional[str],
    run: str,
    in_files: dict,
) -> None:
    """Filter data from a single subject."""
    raise ValueError('Foo')
```

Now I run the pipeline:
```
❯ MNE_BIDS_STUDY_TASK=MMN mne_bids_pipeline --config mne_bids_pipeline/tests/configs/config_ERP_CORE.py
[2022-11-10 11:38:33] ╶╴👋 Welcome aboard the MNE BIDS Pipeline!
[2022-11-10 11:38:33] ╶╴🧾 Using configuration: mne_bids_pipeline/tests/configs/config_ERP_CORE.py
[2022-11-10 11:38:33] ┌╴🚀 scripts/init/_00_init_derivatives_dir Now running  👇
[2022-11-10 11:38:35] └╴🎉 scripts/init/_00_init_derivatives_dir Done running 👆 [2s]
[2022-11-10 11:38:35] ┌╴🚀 scripts/init/_01_find_empty_room Now running  👇
[2022-11-10 11:38:35] │ ⏩ mne_bids_pipeline/_config_import Skipping, empty-room data only relevant for MEG …
[2022-11-10 11:38:35] └╴🎉 scripts/init/_01_find_empty_room Done running 👆 [1s]
[2022-11-10 11:38:35] ┌╴🚀 scripts/preprocessing/_01_maxfilter Now running  👇
[2022-11-10 11:38:35] │ ⏩ preprocessing/_01_maxfilter Skipping …
[2022-11-10 11:38:35] └╴🎉 scripts/preprocessing/_01_maxfilter Done running 👆 [1s]
[2022-11-10 11:38:35] ┌╴🚀 scripts/preprocessing/_02_frequency_filter Now running  👇
[2022-11-10 11:38:35] │ ⏳️ preprocessing/_02_frequency_filter A critical error occurred. The error message was: Foo

Starting post-mortem debugger.
Traceback (most recent call last):
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/_run.py", line 64, in wrapper
    out = memory.cache(func)(*args, **kwargs)
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/_run.py", line 253, in wrapper
    out_files = memorized_func(*args, **kwargs)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/memory.py", line 594, in __call__
    return self._cached_call(args, kwargs)[0]
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/memory.py", line 537, in _cached_call
    out, metadata = self.call(*args, **kwargs)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/memory.py", line 779, in call
    output = self.func(*args, **kwargs)
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/scripts/preprocessing/_02_frequency_filter.py", line 176, in filter_data
    raise ValueError('Foo')
ValueError: Foo
[2022-11-10 11:38:35] │ ⏳️ preprocessing/_02_frequency_filter A critical error occurred. The error message was: Foo

Starting post-mortem debugger.
Traceback (most recent call last):
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/_run.py", line 64, in wrapper
    out = memory.cache(func)(*args, **kwargs)
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/_run.py", line 253, in wrapper
    out_files = memorized_func(*args, **kwargs)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/memory.py", line 594, in __call__
    return self._cached_call(args, kwargs)[0]
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/memory.py", line 537, in _cached_call
    out, metadata = self.call(*args, **kwargs)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/memory.py", line 779, in call
    output = self.func(*args, **kwargs)
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/scripts/preprocessing/_02_frequency_filter.py", line 176, in filter_data
    raise ValueError('Foo')
ValueError: Foo
> /Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/scripts/preprocessing/_02_frequency_filter.py(176)filter_data()
-> raise ValueError('Foo')
> /Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/scripts/preprocessing/_02_frequency_filter.py(176)filter_data()
-> raise ValueError('Foo')
joblib.externals.loky.process_executor._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/_run.py", line 64, in wrapper
    out = memory.cache(func)(*args, **kwargs)
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/_run.py", line 253, in wrapper
    out_files = memorized_func(*args, **kwargs)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/memory.py", line 594, in __call__
    return self._cached_call(args, kwargs)[0]
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/memory.py", line 537, in _cached_call
    out, metadata = self.call(*args, **kwargs)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/memory.py", line 779, in call
    output = self.func(*args, **kwargs)
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/scripts/preprocessing/_02_frequency_filter.py", line 176, in filter_data
    raise ValueError('Foo')
ValueError: Foo

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/externals/loky/process_executor.py", line 436, in _process_worker
    r = call_item()
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/externals/loky/process_executor.py", line 288, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/_parallel_backends.py", line 595, in __call__
    return self.func(*args, **kwargs)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/parallel.py", line 262, in __call__
    return [func(*args, **kwargs)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/parallel.py", line 262, in <listcomp>
    return [func(*args, **kwargs)
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/_run.py", line 109, in wrapper
    pdb.post_mortem(tb)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/pdb.py", line 1642, in post_mortem
    p.interaction(None, t)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/pdb.py", line 357, in interaction
    self._cmdloop()
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/pdb.py", line 322, in _cmdloop
    self.cmdloop()
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/cmd.py", line 126, in cmdloop
    line = input(self.prompt)
RuntimeError: input(): lost sys.stdin
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/hoechenberger/mambaforge/envs/mne/bin/mne_bids_pipeline", line 8, in <module>
    sys.exit(main())
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/run.py", line 171, in main
    script_module.main(config=config_imported)
  File "/Users/hoechenberger/Development/mne-bids-pipeline/mne_bids_pipeline/scripts/preprocessing/_02_frequency_filter.py", line 338, in main
    logs = parallel(
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/parallel.py", line 1056, in __call__
    self.retrieve()
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/parallel.py", line 935, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/site-packages/joblib/_parallel_backends.py", line 542, in wrap_future_result
    return future.result(timeout=timeout)
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/Users/hoechenberger/mambaforge/envs/mne/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
RuntimeError: input(): lost sys.stdin
```
So the debugger doesn't start… or actually, it seems to start but then loses its input terminal (?)

I believe this used to work?

### LINKED ISSUE #655: Feature overview in example gallery is broken
It only says that frequency filter (and only frequency filter) is demonstrated for all examples

https://mne.tools/mne-bids-pipeline/examples/examples.html

### LINKED ISSUE #656: Switch from optparse to argparse
> Deprecated since version 3.2: The [optparse](https://docs.python.org/3/library/optparse.html#module-optparse) module is deprecated and will not be developed further; development will continue with the [argparse](https://docs.python.org/3/library/argparse.html#module-argparse) module.

https://docs.python.org/3/library/optparse.html

### PR DESCRIPTION
**Pull request by larsoner**
Closes #654
Closes #655
Closes #656
Closes #55

Routes more stuff through `_import_config` and through the main `OptParser` which simplifies things a lot I think. Now I get:
```
$ python mne_bids_pipeline/tests/run_tests.py ERP_CORE_ERN --interactive
📝 Running the following tests: ERP_CORE_ERN
[2022-11-10 13:00:55] ╶╴👋 Welcome aboard the MNE BIDS Pipeline!
[2022-11-10 13:00:55] ╶╴🧾 Using configuration: /home/larsoner/python/mne-bids-pipeline/mne_bids_pipeline/tests/configs/config_ERP_CORE.py
[2022-11-10 13:00:55]   ❌ Setting config.on_error="debug" because of interactive mode
[2022-11-10 13:00:55]   ❌ Setting config.N_JOBS=1 because config.on_error="debug"
[2022-11-10 13:00:55]   ❌ Setting config.parallel_backend="loky" because config.on_error="debug"
[2022-11-10 13:00:55] ┌╴🚀 scripts/init/_00_init_derivatives_dir Now running  👇
[2022-11-10 13:00:55] └╴🎉 scripts/init/_00_init_derivatives_dir Done running 👆 [1s]
[2022-11-10 13:00:55] ┌╴🚀 scripts/init/_01_find_empty_room Now running  👇
[2022-11-10 13:00:55] │ ⏩ init/_00_init_derivatives_dir Skipping, empty-room data only relevant for MEG …
[2022-11-10 13:00:55] └╴🎉 scripts/init/_01_find_empty_room Done running 👆 [1s]
[2022-11-10 13:00:55] ┌╴🚀 scripts/preprocessing/_01_maxfilter Now running  👇
[2022-11-10 13:00:55] │ ⏩ preprocessing/_01_maxfilter Skipping …
[2022-11-10 13:00:55] └╴🎉 scripts/preprocessing/_01_maxfilter Done running 👆 [1s]
[2022-11-10 13:00:55] ┌╴🚀 scripts/preprocessing/_02_frequency_filter Now running  👇
[2022-11-10 13:00:55] │ 🐛 preprocessing/_02_frequency_filter A critical error occurred. The error message was: 

Starting post-mortem debugger.
Traceback (most recent call last):
  File "/home/larsoner/python/mne-bids-pipeline/mne_bids_pipeline/_run.py", line 59, in wrapper
    out = memory.cache(func)(*args, **kwargs)
  File "/home/larsoner/python/mne-bids-pipeline/mne_bids_pipeline/_run.py", line 247, in wrapper
    out_files = memorized_func(*args, **kwargs)
  File "/home/larsoner/python/joblib/joblib/memory.py", line 594, in __call__
    return self._cached_call(args, kwargs)[0]
  File "/home/larsoner/python/joblib/joblib/memory.py", line 537, in _cached_call
    out, metadata = self.call(*args, **kwargs)
  File "/home/larsoner/python/joblib/joblib/memory.py", line 795, in call
    output = self.func(*args, **kwargs)
  File "/home/larsoner/python/mne-bids-pipeline/mne_bids_pipeline/scripts/preprocessing/_02_frequency_filter.py", line 179, in filter_data
    raise RuntimeError
RuntimeError

```

### CODE REVIEW

File: mne_bids_pipeline/_main.py [hunk 1]
```diff
@@ -32,46 +35,50 @@ def main():
...
+            bad = True
         else:
-            raise ValueError(bad_msg)
-    elif len(args):
-        raise ValueError(bad_msg)
+            config = config_switch
+    elif config_switch is not None:
+        bad = True
+    if bad:
+        parser.error(bad_msg)
```
Comment (larsoner) [inline mne_bids_pipeline/_main.py:81]: @hoechenberger I replaced the `try/except` stuff with this, I think it's the standard way to handle errors:
```
$ mne_bids_pipeline
usage: mne_bids_pipeline [-h] [--version] [--config FILE] [--steps STEPS] [--root-dir ROOT_DIR] [--subject SUBJECT] [--session SESSION] [--task TASK] [--run RUN]
                         [--n_jobs N_JOBS] [--interactive] [--debug] [--no-cache]
                         [config]
mne_bids_pipeline: error: You must specify a configuration file as a single argument or with --config.
```

Comment (hoechenberger) [reply mne_bids_pipeline/_main.py]: > @hoechenberger I replaced the `try/except` stuff with this, I think it's the standard way to handle errors:

I saw this too when reading the docs, but I thought using our custom logger would be more consistent … but I'm not sure! Could we maybe insert an emoji here, just to get a little bit more colorful output? :)

Comment (larsoner) [reply mne_bids_pipeline/_main.py]: Done
```
mne_bids_pipeline: error: 😵 You must specify a configuration file either as a single argument or with --config, but neither was provided.
```

File: mne_bids_pipeline/_main.py [hunk 2]
```diff
@@ -32,46 +34,48 @@ def main():
...
         else:
-            raise ValueError(bad_msg)
-    elif len(args):
-        raise ValueError(bad_msg)
+            config = config_switch
+    elif config_switch is not None:
+        bad = 'both were provided'
+    if bad:
+        parser.error(
+            '😵 You must specify a configuration file either as a single '
```
Comment (hoechenberger) [inline mne_bids_pipeline/_main.py:77]: Hehe, I love this emoji but I realized in my terminal it's sometimes a little small and I cannot really make out the eyes :( So in #653 I switched from that emoji to a fat red X:
```suggestion
            '❌ You must specify a configuration file either as a single '
```

File: docs/source/changes.md [hunk 3]
```diff
@@ -195,11 +195,11 @@ authors:
...
 - Add time-frequency decoding based on [common spatial patterns](https://mne.tools/stable/auto_examples/decoding/decoding_csp_timefreq.html)
-   (CSP).
-   ({{ gh(625) }} by {{ authors.crsegerie }}, {{ authors.agramfort }},
-    {{ authors.hoechenberger }}, and {{ authors.larsoner }})
+  (CSP).
+  ({{ gh(625) }} by {{ authors.crsegerie }}, {{ authors.agramfort }},
+  {{ authors.hoechenberger }}, and {{ authors.larsoner }})
 - Add progress bar for time-by-time decoding
-  ({{ gh(647) by {{ authors.larsoner }})
+  ({{ gh(647) }} by {{ authors.larsoner }})
```
[the remark this answers is not in the document]
Comment (hoechenberger) [reply docs/source/changes.md:202]: Whoops! Yes!

**Comment by larsoner** [comment]
@hoechenberger okay to mark this one for merge? Happy to help with the rebase/merge of #653 afterward.

Or I can push a test to #653 and mark that one for merge first if you want

**Comment by larsoner** [comment]
Okay, changed the code to use pytest instead of our own custom running functions. This comes with some nice benefits like standardized pytest functionality, but also other stuff becomes easy like:

1. [codecov](https://app.codecov.io/gh/mne-tools/mne-bids-pipeline/pull/658) (spoiler alert: 82%)
2. CircleCI [test info integration](https://app.circleci.com/insights/github/mne-tools/mne-bids-pipeline/workflows/commit/tests?branch=pull%2F658)

Hooray standardized tools! :rocket: 

Ready for review/merge from my end @hoechenberger

**Review by agramfort — APPROVED** [review]

**Comment by agramfort** [comment]
@hoechenberger i let you merge if happy

**Comment by hoechenberger** [comment]
Amazing work, @larsoner

**Pull request merged**